### PR TITLE
VIH-8673 new implementation for gdpr job

### PR DESCRIPTION
**Pull request by omio-ma**
### JIRA link (if applicable) ###
VIH-8673

### Change description ###
- updated the function to use the new endpoints and anonymise using data specified

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesWithSpecifiedDataService.cs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+            await _videoApiClient.AnonymiseConferenceWithHearingIdsAsync(new AnonymiseConferenceWithHearingIdsRequest
+                { HearingIds = anonymisationData.HearingIds });
+            await _videoApiClient.AnonymiseQuickLinkParticipantWithHearingIdsAsync(
+                new AnonymiseQuickLinkParticipantWithHearingIdsRequest { HearingIds = anonymisationData.HearingIds });
+
+            await _bookingsApiClient.AnonymiseParticipantAndCaseByHearingIdAsync("hearingIds",
+                anonymisationData.HearingIds);
+
+            if (anonymisationData.Usernames != null)
```
Comment (oliver-scott) [inline SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesWithSpecifiedDataService.cs:53]: Just a thought, could we invert this if statement to reduce nesting?

```
if (anonymisationData.Usernames == null)
{
  return;
}
            
foreach (var username in anonymisationData.Usernames)
{
  ...
}
```

Comment (omio-ma) [reply SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesWithSpecifiedDataService.cs:53]: yea can do that, however I was wondering if there is something more I can do around this. Can explain why I have done it this way because the whole try catch here isnt ideal I am aware.

File: SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs [hunk 2]
```diff
@@ -12,52 +15,90 @@ public interface IAnonymiseHearingsConferencesDataService
...
+
+                try
+                {
+                    await _bookingsApiClient.AnonymisePersonWithUsernameForExpiredHearingsAsync(username);
+                }
+                catch (BookingsApiException exception)
+                {
+                    _logger.LogError(exception, ProcessingUsernameExceptionMessage,
+                        username);
                 }
```
Comment (will-craig) [inline SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs:92]: As the 'Catch' logic is the same in each, we could maybe tidy it up a bit reduce the duplication and combine the try's, with something like this: 
```
`                try
                {
                    await _videoApiClient.AnonymiseParticipantWithUsernameAsync(username);
                    await _bookingsApiClient.AnonymisePersonWithUsernameForExpiredHearingsAsync(username);
                }
                catch (Exception ex) when (ex is BookingsApiException || ex is VideoApiException)
                {
                    _logger.LogError(ex, ProcessingUsernameExceptionMessage,
                        username);
                }`
```
                
  Pretty sure that syntax is fine, but you may need to double check

File: SchedulerJobs/SchedulerJobs/FeatureFlags.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+﻿namespace SchedulerJobs
+{
+    public static class FeatureFlags
```
Comment (kevin-hmcts) [inline SchedulerJobs/SchedulerJobs/FeatureFlags.cs:3]: Is this file required?

File: SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs [hunk 4]
```diff
@@ -12,52 +15,90 @@ public interface IAnonymiseHearingsConferencesDataService
...
 
-        private async Task DeleteUserAsync()
-        {
-            // delete users from AAD.
-            // get users that do not have hearings in the future and have had hearing more than 3 months in the past. 
-            // (exclude judges, vhos, test users, performance test users.
-            var usersToDelete = await _bookingsApiClient.GetPersonByClosedHearingsAsync();
 
-            if (usersToDelete != null && usersToDelete.Usernames != null)
+            if (anonymisationData.Usernames == null) return;
```
Comment (kevin-hmcts) [inline SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs:57]: Is it worth a log here to indicate there was nothing to process

File: SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs [hunk 5]
```diff
@@ -12,52 +15,90 @@ public interface IAnonymiseHearingsConferencesDataService
...
-                        if (exception.StatusCode != (int)HttpStatusCode.NotFound && exception.StatusCode != (int)HttpStatusCode.Forbidden)
-                        {
-                            throw;
-                        }
+                        _logger.LogError(exception, ProcessingUsernameExceptionMessage,
+                            username);
                     }
+
+                try
+                {
```
Comment (kevin-hmcts) [inline SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs:75]: Can these two try..catch be a single try..catch with a switch on the exception type

File: SchedulerJobs/SchedulerJobs/FeatureFlags.cs (hunk 3 above)
Comment (omio-ma) [reply SchedulerJobs/SchedulerJobs/FeatureFlags.cs:3]: no this can be removed

File: SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs (hunk 2 above)
Comment (omio-ma) [reply SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs:92]: I did consider this, however the problem with this is, if an exception gets thrown at video api, it wont make the request to bookings api anymore, thats why I left it as seperate try catch instead

File: SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs (hunk 5 above)
Comment (omio-ma) [reply SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs:75]: as I have mentioned this in the earlier review comment, this would mean the whole process will stop if any one of the components throw an error and I thought the idea was we would like it to continue to process other endpoints even if one of them throws an error?

File: SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs (hunk 4 above)
Comment (omio-ma) [reply SchedulerJobs/SchedulerJobs.Services/AnonymiseHearingsConferencesDataService.cs:57]: I dont think we need to log this, especially because we are also logging all the hearing ids being processed earlier which should give enough information about what is being processed.
